src/sinks/gcp/chronicle_unstructured.rs

```diff
@@ -0,0 +1,546 @@
+
+        let request_settings = RequestSettings::new(self)?;
+
+        let sink = GcsSink::new(svc, request_settings, partitioner, batch_settings);
+
+        Ok(VectorSink::from_event_streamsink(sink))
+    }
+
+    fn key_partitioner(&self) -> crate::Result<KeyPartitioner> {
+        Ok(KeyPartitioner::new(self.log_type.clone()))
+    }
+
+    fn create_endpoint(&self, path: &str) -> Result<String, ChronicleError> {
+        Ok(format!(
+            "{}/{}",
+            match (&self.endpoint, self.region) {
+                (Some(endpoint), None) => endpoint.trim_end_matches('/'),
+                (None, Some(region)) => region.endpoint(),
+                (Some(_), Some(_)) => return Err(ChronicleError::BothRegionAndEndpoint),
+                (None, None) => return Err(ChronicleError::RegionOrEndpoint),
+            },
+            path
+        ))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ChronicleRequest {
+    pub body: Bytes,
+    pub finalizers: EventFinalizers,
+    pub metadata: RequestMetadata,
+}
+
+impl Finalizable for ChronicleRequest {
+    fn take_finalizers(&mut self) -> EventFinalizers {
+        std::mem::take(&mut self.finalizers)
+    }
+}
+
+#[derive(Clone, Debug)]
+struct ChronicleEncoder {
+    customer_id: String,
+    encoder: codecs::Encoder<()>,
+    transformer: codecs::Transformer,
+}
+
+impl Encoder<(String, Vec<Event>)> for ChronicleEncoder {
+    fn encode_input(
+        &self,
+        input: (String, Vec<Event>),
+        writer: &mut dyn io::Write,
+    ) -> io::Result<usize> {
+        let (partition_key, events) = input;
+        let mut encoder = self.encoder.clone();
+        let events = events
+            .into_iter()
+            .filter_map(|mut event| {
+                let timestamp = event
+                    .as_log()
+                    .get(log_schema().timestamp_key())
+                    .and_then(|ts| ts.as_timestamp())
+                    .cloned();
+                let mut bytes = BytesMut::new();
+                self.transformer.transform(&mut event);
+                encoder.encode(event, &mut bytes).ok()?;
+
+                let mut value = json!({
+                    "log_text": String::from_utf8_lossy(&bytes),
+                });
+
+                if let Some(ts) = timestamp {
+                    value.as_object_mut().unwrap().insert(
+                        "ts_rfc3339".to_string(),
+                        ts.to_rfc3339_opts(chrono::SecondsFormat::AutoSi, true)
+                            .into(),
+                    );
+                }
+
+                Some(value)
+            })
+            .collect::<Vec<_>>();
+
+        let json = json!({
+            "customer_id": self.customer_id,
+            "log_type": partition_key,
+            "entries": events,
+        });
+
+        let body = crate::serde::json::to_bytes(&json)?.freeze();
+        writer.write_all(&body)?;
+        Ok(body.len())
```

Comment:
We can remove the temporary allocation here by following the same pattern as in https://github.com/vectordotdev/vector/blob/32022dbff7be1fd3f92ade125feca2878e27e941/src/sinks/new_relic/encoding.rs#L21-L25

```suggestion
    let size = as_tracked_write::<_, _, io::Error>(writer, &json, |writer, json| {
        writer.write_all(json)?;
        Ok(())
    })?;
    io::Result::Ok(size)
```